File: java-union-reader-writer-example/src/main/java/com/intellicess/union/java_union_reader_writer_example/client/UnionClient.java
package com.intellicess.union.java_union_reader_writer_example.client;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.intellicess.union.java_union_reader_writer_example.client.model.*;
import com.intellicess.union.java_union_reader_writer_example.client.model.auth.AccessTokenResponse;

import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.LocalDateTime;
import java.time.temporal.ChronoUnit;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

public class UnionClient {

    private static final long MAX_ACCESS_TOKEN_AGE_SECONDS = 60;
    private static final int MAX_NUMBER_OF_IDS_PER_REQUEST = 25;
    private static final String APPLICATION_X_WWW_FORM_URLENCODED_CONTENT_TYPE = "application/x-www-form-urlencoded";
    private static final String CONTENT_TYPE_HEADER_NAME = "Content-Type";
    public static final String ACCESS_TOKEN_BEARER_PREFIX = "Bearer ";

    private final HttpClient httpClient;
    private final ObjectMapper objectMapper;
    private final UnionClientConfiguration config;

    private String lastAccessToken;
    private LocalDateTime accessTokenCreationTime;
    private final Map<String, Long> searchSinceTimestampMap = new ConcurrentHashMap<>();

    public UnionClient(UnionClientConfiguration unionClientConfiguration, ObjectMapper objectMapper) {
        this.httpClient = HttpClient.newHttpClient();
        this.objectMapper = objectMapper;
        this.config = unionClientConfiguration;
    }

    public JwlfSavedResponse saveJwlfLog(String client, String region, String asset, String folder, JwlfRequest jwlfLog) {
        String json = toJson(jwlfLog);
        String url = config.getUnionUrl() + "/api/v1/well-logs/" + client + "/" + region + "/" + asset + "/" + folder;
        String accessToken = getAccessToken();
        HttpRequest request = HttpRequest
                .newBuilder()
                .uri(URI.create(url))
                .header("Authorization", ACCESS_TOKEN_BEARER_PREFIX + accessToken)
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(json))
                .build();

        HttpResponse<String> response = sendRequest(request, HttpResponse.BodyHandlers.ofString());
        String responseBody = response.body();
        if (response.statusCode() >= 300) {
            throw new RuntimeException("UnionClient#saveJwlfLog failure with http response status code=" + response.statusCode()
                    + " and response=" + responseBody);
        }
        return fromJson(responseBody, JwlfSavedResponse.class);
    }

    public JwlfSavedResponseList saveJwlfLogs(String client, String region, String asset, String folder,
                                              List<JwlfRequest> jwlfLogs) {
        String json = toJson(jwlfLogs);
        String url = config.getUnionUrl() + "/api/v1/well-logs/" + client + "/" + region + "/" + asset + "/" + folder
                + "/batch-save";
        String accessToken = getAccessToken();
        HttpRequest request = HttpRequest
                .newBuilder()
                .uri(URI.create(url))
                .header("Authorization", "Bearer " + accessToken)
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(json))
                .build();
        HttpResponse<String> response = sendRequest(request, HttpResponse.BodyHandlers.ofString());
        String responseBody = response.body();
        if (response.statusCode() >= 300) {
            throw new RuntimeException("UnionClient#saveJwlfLogs failure with http response status code="
                    + response.statusCode() + " and response=" + responseBody);
        }
        return fromJson(responseBody, JwlfSavedResponseList.class);

    }

    public JwlfResponseList getNewJwlfLogsWithData(String client, String region, String asset, String folder,
                                                   Long inclusiveSinceTimestamp) {
        String searchKey = client + "/" + region + "/" + asset + "/" + folder;
        Long currentInclusiveSinceTimestamp = searchSinceTimestampMap.getOrDefault(searchKey, inclusiveSinceTimestamp);

        long stableDataTimestamp = getStableJwlfDataTimestamp(client, region, asset, folder);
        long tillTimestampExclusive = stableDataTimestamp + 1;

        JwlfResponseList jwlfLogsWithoutData = getJwlfLogsWithoutData(client, region, asset, folder,
                currentInclusiveSinceTimestamp, tillTimestampExclusive);
        List<JwlfResponse> logsWithoutData = jwlfLogsWithoutData.getList();
        List<String> ids = logsWithoutData.stream()
                .map(JwlfResponse::getId)
                .collect(Collectors.toList());

        searchSinceTimestampMap.put(searchKey, tillTimestampExclusive);
        if (ids.size() == 0) {
            JwlfResponseList jwlfResponseList = new JwlfResponseList();
            jwlfResponseList.setList(Collections.emptyList());
            jwlfResponseList.setStableDataTimestamp(jwlfLogsWithoutData.getStableDataTimestamp());
            return jwlfResponseList;
        }
        return getJwlfLogsWithData(client, region, asset, folder, ids);
    }

    public JwlfResponseList getJwlfLogsWithData(String client, String region, String asset, String folder,
                                                Collection<String> ids) {
        List<List<String>> idsBatches = new ArrayList<>();
        idsBatches.add(new ArrayList<>());
        int idBatchIndex = 0;
        int idIndexInBatch = 0;
        for (String jwlfId : ids) {
            if (idIndexInBatch >= MAX_NUMBER_OF_IDS_PER_REQUEST) {
                idsBatches.add(new ArrayList<>());
                idBatchIndex++;
                idIndexInBatch = 0;
            }
            List<String> idsBatch = idsBatches.get(idBatchIndex);
            idsBatch.add(jwlfId);
            idIndexInBatch++;
        }

        String baseUrl = config.getUnionUrl() + "/api/v1/well-logs/" + client + "/" + region + "/" + asset + "/" + folder;
        List<JwlfResponse> jwlfLogs = new ArrayList<>();
        long stableDataTimestamp = 0L;
        for (List<String> idsBatch : idsBatches) {
            String url = baseUrl + "?" + mapToUrlEncodedForm(Map.of("id", idsBatch));
            String accessToken = getAccessToken();
            HttpRequest request = HttpRequest
                    .newBuilder()
                    .uri(URI.create(url))
                    .header("Authorization", "Bearer " + accessToken)
                    .GET()
                    .build();
            HttpResponse<String> response = sendRequest(request, HttpResponse.BodyHandlers.ofString());
            String responseBody = response.body();
            if (response.statusCode() >= 300) {
                throw new RuntimeException("UnionClient#getJwlfLogsWithData failure with http response status code="
                        + response.statusCode() + " and response=" + responseBody);
            }
            JwlfResponseList jwlfResponseList = fromJson(responseBody, JwlfResponseList.class);
            jwlfLogs.addAll(jwlfResponseList.getList());
            stableDataTimestamp = jwlfResponseList.getStableDataTimestamp();
        }
        JwlfResponseList jwlfResponseList = new JwlfResponseList();
        jwlfResponseList.setList(jwlfLogs);
        jwlfResponseList.setStableDataTimestamp(stableDataTimestamp);
        return jwlfResponseList;
    }

    public JwlfResponseList getJwlfLogsWithoutData(String client, String region, String asset, String folder,
                                                   Long sinceTimestampInclusive, Long tillTimestampExclusive) {
        String baseUrl = config.getUnionUrl() + "/api/v1/well-logs/" + client + "/" + region + "/" + asset + "/" + folder;

        Map<String, ?> queryParams = Map.of("sinceTimestamp", sinceTimestampInclusive, "tillTimestamp", tillTimestampExclusive);
        String url = baseUrl + "?" + mapToUrlEncodedForm(queryParams);
        String accessToken = getAccessToken();
        HttpRequest request = HttpRequest
                .newBuilder()
                .uri(URI.create(url))
                .header("Authorization", "Bearer " + accessToken)
                .GET()
                .build();
        HttpResponse<String> response;
        response = sendRequest(request, HttpResponse.BodyHandlers.ofString());
        String responseBody = response.body();
        if (response.statusCode() >= 300) {
            throw new RuntimeException("UnionClient#getJwlfLogsWithoutData failure with http response status code="
                    + response.statusCode() + " and response=" + responseBody);
        }
        return fromJson(responseBody, JwlfResponseList.class);
    }

    public long getStableJwlfDataTimestamp(String client, String region, String asset, String folder) {
        try {
            JwlfResponseList jwlfLogsWithoutData = getJwlfLogsWithoutData(client, region, asset, folder, 0L, 0L);
            return jwlfLogsWithoutData.getStableDataTimestamp();
        } catch (RuntimeException e) {
            throw new RuntimeException("UnionClient#getStableJwlfDataTimestamp failure caused by the following exception", e);
        }
    }

    private String getAccessToken() {
        long accessTokenAgeSeconds = Optional.ofNullable(accessTokenCreationTime)
                .map(creationTime -> ChronoUnit.SECONDS.between(creationTime, LocalDateTime.now()))
                .orElse(MAX_ACCESS_TOKEN_AGE_SECONDS);
        if (accessTokenAgeSeconds < MAX_ACCESS_TOKEN_AGE_SECONDS) {
            return lastAccessToken;
        }
        String url = config.getAuthProviderUrl() + "/auth/realms/" + config.getAuthProviderRealm() + "/protocol/openid-connect/token";
        String urlEncodedForm = mapToUrlEncodedForm(Map.of(
                // "client_secret", clientSecretIfSetupInAuthProviderForClient,
                // "totp", totpFromUserPhoneIf2FAIsSetup,
                "client_id", config.getAuthClientId(),
                "grant_type", "password",
                "username", config.getUsername(),
                "password", config.getPassword()
        ));
        HttpRequest request = HttpRequest
                .newBuilder()
                .uri(URI.create(url))
                .header(CONTENT_TYPE_HEADER_NAME, APPLICATION_X_WWW_FORM_URLENCODED_CONTENT_TYPE)
                .POST(HttpRequest.BodyPublishers.ofString(urlEncodedForm))
                .build();
        HttpResponse<String> response = sendRequest(request, HttpResponse.BodyHandlers.ofString());
        String responseBody = response.body();
        if (response.statusCode() >= 300) {
            throw new RuntimeException("UnionClient#getAccessToken failure with http response status code="
                    + response.statusCode() + " and response=" + responseBody);
        }
        AccessTokenResponse accessTokenResponse = fromJson(responseBody, AccessTokenResponse.class);
        lastAccessToken = accessTokenResponse.getAccessToken();
        accessTokenCreationTime = LocalDateTime.now();
        return lastAccessToken;
    }

    private <T> HttpResponse<T> sendRequest(HttpRequest request, HttpResponse.BodyHandler<T> responseBodyHandler) {
        HttpResponse<T> response;
        try {
            response = httpClient.send(request, responseBodyHandler);
        } catch (IOException | InterruptedException e) {
            if (e.getMessage().contains("GOAWAY")) {
                try {
                    response = httpClient.send(request, responseBodyHandler);
                } catch (IOException | InterruptedException ex) {
                    throw new RuntimeException(ex);
                }
            } else {
                throw new RuntimeException(e);
            }
        }
        return response;
    }

    private String mapToUrlEncodedForm(Map<String, ?> map) {
        return map.entrySet()
                .stream()
                .filter(entry -> entry.getValue() != null)
                .map(this::mapEntryToUrlEncodedForm)
                .flatMap(Collection::stream)
                .collect(Collectors.joining("&"));
    }

    private List<String> mapEntryToUrlEncodedForm(Map.Entry<String, ?> entry) {
        if (entry.getValue() instanceof Collection) {
            return (List<String>)
                    ((Collection) entry.getValue()).stream()
                            .filter(Objects::nonNull)
                            .map(valueElement -> mapEntryToUrlEncodedForm(Map.entry(entry.getKey(), valueElement)))
                            .flatMap(o -> ((List<String>) o).stream())
                            .collect(Collectors.toList());
        }
        String stringValue = String.valueOf(entry.getValue());
        return List.of(entry.getKey() + "=" + URLEncoder.encode(stringValue, StandardCharsets.UTF_8));
    }

    private String toJson(Object object) {
        try {
            return objectMapper.writeValueAsString(object);
        } catch (JsonProcessingException e) {
            throw new RuntimeException(e);
        }
    }

    private <T> T fromJson(String json, Class<T> clazz) {
        try {
            return objectMapper.readValue(json, clazz);
        } catch (JsonProcessingException e) {
            throw new RuntimeException(e);
        }
    }
}

File: csharp-union-reader-writer-example/csharp-union-reader-writer-example/UnionClient.cs
namespace Union;

using Union;
using System;
using Serilog.Core;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Net.Http.Json;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Web;

public class UnionClient
{

    private const ulong MAX_ACCESS_TOKEN_AGE_SECONDS = 60;
    private const int MAX_NUMBER_OF_IDS_PER_REQUEST = 25;
    private const string APPLICATION_X_WWW_FORM_URLENCODED_CONTENT_TYPE = "application/x-www-form-urlencoded";
    private const string ACCESS_TOKEN_BEARER_PREFIX = "Bearer ";
	private static Logger log = AppLogFactory.CreateLogger();

    private string _lastAccessToken;
    private DateTime? _accessTokenCreationTime = null;
    private ConcurrentDictionary<string, long> _searchSinceTimestampDict = new ConcurrentDictionary<string, long>();
    
    private readonly UnionClientConfiguration _config;
    private readonly HttpClient _httpClient;


    public UnionClient(UnionClientConfiguration config)
    {
        _config = config;
        _httpClient = new HttpClient();
    }

    public JwlfSavedResponse SaveJwlfLog(string client, string region, string asset, string folder, JwlfRequest jwlfLog)
    {
        string url = _config.UnionUrl + "/api/v1/well-logs/" + client + "/" + region + "/" + asset + "/" + folder;
        string accessToken = GetAccessToken();

        var options = new JsonSerializerOptions(JsonSerializerDefaults.Web);
        options.Converters.Add(new JwlfValueTypeJsonConverter());

        var jsonContent =  JsonContent.Create(jwlfLog, mediaType: null, options);
        _httpClient.DefaultRequestHeaders.Authorization = AuthenticationHeaderValue.Parse(ACCESS_TOKEN_BEARER_PREFIX + accessToken);
        var response =  _httpClient.PostAsync(url, jsonContent).Result;
        var content = response.Content;
        if ((int)response.StatusCode >= 300)
        {
            throw new Exception("UnionClient#SaveJwlfLog failure with http response status code=" + response.StatusCode
                    + " and response=" + content.ReadAsStringAsync().Result);
        }
        JwlfSavedResponse responseBody = content.ReadFromJsonAsync<JwlfSavedResponse>(options).Result;
        return responseBody;
    }

    public JwlfSavedResponseList SaveJwlfLogs(string client, string region, string asset, string folder, List<JwlfRequest> jwlfLogs)
    {
        string url = _config.UnionUrl + "/api/v1/well-logs/" + client + "/" + region + "/" + asset + "/" + folder + "/batch-save";
        string accessToken = GetAccessToken();

        var options = new JsonSerializerOptions(JsonSerializerDefaults.Web);
        options.Converters.Add(new JwlfValueTypeJsonConverter());
        
        var jsonContent =  JsonContent.Create(jwlfLogs, mediaType: null, options);
		log.Information("sending jsonContent={jsonContent}", jsonContent.ToString());
        _httpClient.DefaultRequestHeaders.Authorization = AuthenticationHeaderValue.Parse(ACCESS_TOKEN_BEARER_PREFIX + accessToken);
        var response =  _httpClient.PostAsync(url, jsonContent).Result;
        var content = response.Content;
        if ((int)response.StatusCode >= 300)
        {
            throw new Exception("UnionClient#SaveJwlfLogs failure with http response status code=" + response.StatusCode
                    + " and response=" + content.ToString());
        }
        JwlfSavedResponseList responseBody = content.ReadFromJsonAsync<JwlfSavedResponseList>(options).Result;
        return responseBody;
    }

    public JwlfResponseList GetNewJwlfLogsWithData(string client, string region, string asset, string folder,
                                                        long? inclusiveSinceTimestamp)
    {
        string searchKey = client + "/" + region + "/" + asset + "/" + folder;
        long inclusiveSinceTimestampValue = inclusiveSinceTimestamp.GetValueOrDefault(0);
        long currentInclusiveSinceTimestamp = _searchSinceTimestampDict.GetValueOrDefault(searchKey, inclusiveSinceTimestampValue);

        long stableDataTimestamp = GetStableJwlfDataTimestamp(client, region, asset, folder);
        long tillTimestampExclusive = stableDataTimestamp + 1;

        JwlfResponseList jwlfLogsWithoutData = GetJwlfLogsWithoutData(client, region, asset, folder,
                currentInclusiveSinceTimestamp, tillTimestampExclusive);
        List<JwlfResponse> logsWithoutData = jwlfLogsWithoutData.List;
        List<string> ids = logsWithoutData.Select(jwlfLog => jwlfLog.Id).ToList();

        _searchSinceTimestampDict[searchKey] = tillTimestampExclusive;
        if (ids.Count() == 0) {
            return new JwlfResponseList
            {
                List = new List<JwlfResponse>(),
                StableDataTimestamp = jwlfLogsWithoutData.StableDataTimestamp
            };
        }
        return GetJwlfLogsWithData(client, region, asset, folder, ids);
    }

    public JwlfResponseList GetJwlfLogsWithData(string client, string region, string asset, string folder,
                                                List<string> ids) {
        List<List<string>> idsBatches = new List<List<string>>();
        idsBatches.Add(new List<string>());
        int idBatchIndex = 0;
        int idIndexInBatch = 0;
        foreach (string jwlfId in ids) {
            if (idIndexInBatch >= MAX_NUMBER_OF_IDS_PER_REQUEST) {
                idsBatches.Add(new List<string>());
                idBatchIndex++;
                idIndexInBatch = 0;
            }
            List<string> idsBatch = idsBatches[idBatchIndex];
            idsBatch.Add(jwlfId);
            idIndexInBatch++;
        }

        string baseUrl = _config.UnionUrl + "/api/v1/well-logs/" + client + "/" + region + "/" + asset + "/" + folder;
        List<JwlfResponse> jwlfLogs = new List<JwlfResponse>();
        long stableDataTimestamp = 0;
        foreach (List<string> idsBatch in idsBatches) {
            string url = baseUrl + "?" + string.Join("&", idsBatch.Select(id => "id=" + HttpUtility.UrlEncode(id)));
            string accessToken = GetAccessToken();
            var options = new JsonSerializerOptions(JsonSerializerDefaults.Web);
            options.Converters.Add(new JwlfValueTypeJsonConverter());
            options.Converters.Add(new ObjectDeserializer());

            _httpClient.DefaultRequestHeaders.Authorization = AuthenticationHeaderValue.Parse(ACCESS_TOKEN_BEARER_PREFIX + accessToken);
            var response = _httpClient.GetAsync(url).Result;
            var content = response.Content;
            if ((int)response.StatusCode >= 300)
            {
                throw new Exception("UnionClient#GetJwlfLogsWithData failure with http response status code=" + response.StatusCode
                        + " and response=" + content.ReadAsStringAsync().Result);
            }
            JwlfResponseList jwlfResponseList = content.ReadFromJsonAsync<JwlfResponseList>(options).Result;
            jwlfLogs.AddRange(jwlfResponseList.List);
            stableDataTimestamp = jwlfResponseList.StableDataTimestamp;
        }
        return new JwlfResponseList
        {
            List = jwlfLogs,
            StableDataTimestamp = stableDataTimestamp
        };
    }
    
    public JwlfResponseList GetJwlfLogsWithoutData(String client, String region, String asset, String folder,
                                                   long sinceTimestampInclusive, long tillTimestampExclusive) {
        string baseUrl = _config.UnionUrl + "/api/v1/well-logs/" + client + "/" + region + "/" + asset + "/" + folder;
        string sinceTimestampInclusiveUrlEncoded = HttpUtility.UrlEncode(sinceTimestampInclusive.ToString());
        string url = baseUrl + "?sinceTimestamp=" + sinceTimestampInclusiveUrlEncoded + "&tillTimestamp=" + HttpUtility.UrlEncode(tillTimestampExclusive.ToString());
        string accessToken = GetAccessToken();
        var options = new JsonSerializerOptions(JsonSerializerDefaults.Web);
        options.Converters.Add(new JwlfValueTypeJsonConverter());
        options.Converters.Add(new ObjectDeserializer());

        _httpClient.DefaultRequestHeaders.Authorization = AuthenticationHeaderValue.Parse(ACCESS_TOKEN_BEARER_PREFIX + accessToken);
        var response =  _httpClient.GetAsync(url).Result;
        var content = response.Content;
        if ((int)response.StatusCode >= 300)
        {
            throw new Exception("UnionClient#GetJwlfLogsWithoutData failure with http response status code=" + response.StatusCode
                    + " and response=" + content.ReadAsStringAsync().Result);
        }
        return content.ReadFromJsonAsync<JwlfResponseList>(options).Result;
    }

    public long GetStableJwlfDataTimestamp(string client, string region, string asset, string folder) {
        try {
            JwlfResponseList jwlfLogsWithoutData = GetJwlfLogsWithoutData(client, region, asset, folder, 0, 0);
            return jwlfLogsWithoutData.StableDataTimestamp;
        } catch (Exception e) {
            throw new Exception("UnionClient#getStableJwlfDataTimestamp failure caused by the following exception", e);
        }
    }

    private string GetAccessToken()
    {
        ulong accessTokenAgeSeconds = MAX_ACCESS_TOKEN_AGE_SECONDS;
        if (_accessTokenCreationTime.HasValue)
        {
            accessTokenAgeSeconds = (ulong)((DateTime.Now - _accessTokenCreationTime.Value).TotalSeconds);
        }
        if (accessTokenAgeSeconds < MAX_ACCESS_TOKEN_AGE_SECONDS) {
            return _lastAccessToken;
        }
        var form = new Dictionary<string, string>
        {
            { "client_id", _config.AuthClientId },
            { "grant_type", "password" },
            { "username", _config.Username },
            { "password", _config.Password },
        };
        string url = _config.AuthProviderUrl + "/auth/realms/" + _config.AuthProviderRealm + "/protocol/openid-connect/token";

        var response = _httpClient.PostAsync(url, new FormUrlEncodedContent(form)).Result;
        var content = response.Content;
        if ((int)response.StatusCode >= 300)
        {
            throw new Exception("UnionClient#GetAccessToken failure with http response status code="
                    + response.StatusCode + " and response=" + content.ToString());
        }
        var options = new JsonSerializerOptions(JsonSerializerDefaults.Web);
        AccessTokenResponse accessTokenResponse = content.ReadFromJsonAsync<AccessTokenResponse>(options).Result;
        _lastAccessToken = accessTokenResponse.AccessToken;
        _accessTokenCreationTime = DateTime.Now;
        return _lastAccessToken;
    }
}
